File: searchlib/src/tests/queryeval/predicate/predicate_search_test.cpp
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
// Unit tests for predicate_search.

#include <vespa/searchlib/fef/termfieldmatchdata.h>
#include <vespa/searchlib/fef/termfieldmatchdataarray.h>
#include <vespa/searchlib/queryeval/predicate_search.h>
#include <vespa/vespalib/gtest/gtest.h>
#include <vespa/vespalib/util/arraysize.h>

#include <vespa/log/log.h>
LOG_SETUP("predicate_search_test");

using search::fef::TermFieldMatchData;
using search::fef::TermFieldMatchDataArray;
using namespace search::queryeval;
using namespace search::predicate;
using std::pair;
using std::vector;
using vespalib::arraysize;

namespace {

class MyPostingList : public PredicatePostingList {
    vector<pair<uint32_t, uint32_t>> _entries;
    size_t _index;
    uint32_t _interval;

    void setInterval(uint32_t interval) { _interval = interval; }

public:
    MyPostingList(const vector<pair<uint32_t, uint32_t>> &entries)
        : _entries(entries),
          _index(0),
          _interval(0)
    {
    }
    MyPostingList(std::initializer_list<pair<uint32_t, uint32_t>> ilist)
        : _entries(ilist.begin(), ilist.end()),
          _index(0),
          _interval(0)
    {
    }

    ~MyPostingList() override;

    bool next(uint32_t doc_id) override {
        if (_index < _entries.size()) {
            while (_entries[_index].first <= doc_id) {
                ++_index;
                if (_index == _entries.size()) {
                    setDocId(search::endDocId);
                    return false;
                }
            }
            setDocId(_entries[_index].first);
            setInterval(_entries[_index].second);
            return true;
        }
        setDocId(search::endDocId);
        return false;
    }

    bool nextInterval() override {
        if (_index + 1 < _entries.size() &&
            _entries[_index].first == _entries[_index + 1].first) {
            ++_index;
            setInterval(_entries[_index].second);
            return true;
        }
        return false;
    }
    uint32_t getInterval() const override { return _interval; }
};

MyPostingList::~MyPostingList() = default;

template <int N>
vector<PredicatePostingList::UP>
make_posting_lists_vector(MyPostingList (&plists)[N]) {
    vector<PredicatePostingList::UP> posting_lists;
    for (int i = 0; i < N; ++i) {
        posting_lists.emplace_back(std::make_unique<MyPostingList>(plists[i]));
    }
    return posting_lists;
}

TermFieldMatchDataArray tfmda;
using CV = std::vector<uint8_t>;
using MF = std::vector<uint8_t>;
using IR = std::vector<uint16_t>;

TEST(PredicateSearchTest, Require_that_the_skipping_is_efficient) {
    const uint8_t min_feature[] = { 7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,
                                    7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7};
    const uint8_t kv[] = { 6,7,6,7,6,7,6,8,6,5,6,7,6,0,6,7,
                           7,6,7,6,6,6,6,7,7,7,8,7,8,7,7,7,6,7};
    SkipMinFeature::UP skip = SkipMinFeature::create(min_feature, kv, 34);
    EXPECT_EQ(1u, skip->next());
    EXPECT_EQ(3u, skip->next());
    EXPECT_EQ(5u, skip->next());
    EXPECT_EQ(7u, skip->next());
    EXPECT_EQ(11u, skip->next());
    EXPECT_EQ(15u, skip->next());
    EXPECT_EQ(16u, skip->next());
    EXPECT_EQ(18u, skip->next());
    EXPECT_EQ(23u, skip->next());
    EXPECT_EQ(24u, skip->next());
    EXPECT_EQ(25u, skip->next());
    EXPECT_EQ(26u, skip->next());
    EXPECT_EQ(27u, skip->next());
    EXPECT_EQ(28u, skip->next());
    EXPECT_EQ(29u, skip->next());
    EXPECT_EQ(30u, skip->next());
    EXPECT_EQ(31u, skip->next());
    EXPECT_EQ(33u, skip->next());
}

TEST(PredicateSearchTest, require_that_empty_search_yields_no_results) {
    vector<PredicatePostingList::UP> posting_lists;
    MF mf(3); CV cv(3); IR ir(3, 0xffff);
    PredicateSearch search(&mf[0], &ir[0], 0xffff, cv, std::move(posting_lists), tfmda);
    search.initFullRange();
    EXPECT_EQ(SearchIterator::beginId(), search.getDocId());
    EXPECT_FALSE(search.seek(2));
    EXPECT_TRUE(search.isAtEnd());
}

TEST(PredicateSearchTest, require_that_simple_search_yields_result) {
    MyPostingList plists[] = {{{2, 0x0001ffff}}};
    MF mf{0, 0, 0};
    CV cv{0, 0, 1};
    IR ir(3, 0xffff);
    PredicateSearch search(&mf[0], &ir[0], 0xffff, cv, make_posting_lists_vector(plists), tfmda);
    search.initFullRange();
    EXPECT_EQ(SearchIterator::beginId(), search.getDocId());
    EXPECT_FALSE(search.seek(1));
    EXPECT_EQ(2u, search.getDocId());
    EXPECT_TRUE(search.seek(2));
    EXPECT_EQ(2u, search.getDocId());
    EXPECT_FALSE(search.seek(3));
    EXPECT_TRUE(search.isAtEnd());
}

TEST(PredicateSearchTest, require_that_minFeature_K_is_used_to_prune_results) {
    MyPostingList plists[] = {{{2, 0x0001ffff}},
                              {{5, 0x0001ffff}}};
    MF mf{0, 0, 3, 0, 0, 0};
    CV cv{1, 0, 0, 0, 0, 1};
    IR ir(6, 0xffff);
    PredicateSearch search(&mf[0], &ir[0], 0xffff, cv, make_posting_lists_vector(plists), tfmda);
    search.initFullRange();
    EXPECT_FALSE(search.seek(2));
    EXPECT_EQ(5u, search.getDocId());
}

TEST(PredicateSearchTest, require_that_a_high_K_min_feature_minus_1_can_yield_results) {
    MyPostingList plists[] = {{{2, 0x00010001}},
                              {{2, 0x0002ffff}}};
    MF mf{0, 0, 2};
    CV cv{0, 0, 2};
    IR ir(3, 0xffff);
    PredicateSearch search(&mf[0], &ir[0], 0xffff, cv, make_posting_lists_vector(plists), tfmda);
    search.initFullRange();
    EXPECT_TRUE(search.seek(2));
}

TEST(PredicateSearchTest, require_that_we_can_skip_past_entries) {
    MyPostingList plists[] = {{{2, 0x0001ffff},
                               {5, 0x0001ffff}}};
    MF mf{0, 0, 0, 0, 0, 0};
    CV cv{0, 0, 1, 0, 0, 1};
    IR ir(6, 0xffff);
    PredicateSearch search(&mf[0], &ir[0], 0xffff, cv, make_posting_lists_vector(plists), tfmda);
    search.initFullRange();
    EXPECT_TRUE(search.seek(5));
}

TEST(PredicateSearchTest, require_that_posting_lists_are_sorted_after_advancing) {
    MyPostingList plists[] = {{{1, 0x0001ffff},
                               {5, 0x0001ffff}},
                              {{2, 0x0001ffff},
                               {4, 0x0001ffff}}};
    MF mf{0, 2, 0, 0, 0, 0};
    CV cv{0, 1, 1, 0, 1, 1};
    IR ir(6, 0xffff);
    PredicateSearch search(&mf[0], &ir[0], 0xffff, cv, make_posting_lists_vector(plists), tfmda);
    search.initFullRange();
    EXPECT_FALSE(search.seek(1));
    EXPECT_FALSE(search.seek(3));
    EXPECT_TRUE(search.seek(4));
}

TEST(PredicateSearchTest, require_that_short_interval_ranges_works) {
    MyPostingList plists[] = {{{1, 0x00010001},
                               {5, 0x00010001}},
                              {{2, 0x00010001},
                               {4, 0x00010001}}};
    MF mf{0, 2, 0, 0, 0, 0};
    CV cv{0, 1, 1, 0, 1, 1};
    IR ir(6, 0x0001);
    PredicateSearch search(&mf[0], &ir[0], 0x1, cv, make_posting_lists_vector(plists), tfmda);
    search.initFullRange();
    EXPECT_FALSE(search.seek(1));
    EXPECT_FALSE(search.seek(3));
    EXPECT_TRUE(search.seek(4));
}

TEST(PredicateSearchTest, require_that_empty_posting_lists_work) {
    MyPostingList plists[] = {{}};
    MF mf(3); CV cv(3); IR ir(3, 0xffff);
    PredicateSearch search(&mf[0], &ir[0], 0xffff, cv, make_posting_lists_vector(plists), tfmda);
    search.initFullRange();
    EXPECT_EQ(SearchIterator::beginId(), search.getDocId());
    EXPECT_FALSE(search.seek(2));
    EXPECT_TRUE(search.isAtEnd());
}

TEST(PredicateSearchTest, require_that_shorter_posting_list_ending_is_ok) {
    MyPostingList plists[] = {{{1, 0x0001ffff},
                               {2, 0x0001ffff}},
                              {{4, 0x0001ffff}}};
    MF mf{0, 0, 0, 0, 0};
    CV cv{0, 1, 1, 0, 1};
    IR ir(5, 0xffff);
    PredicateSearch search(&mf[0], &ir[0], 0xffff, cv, make_posting_lists_vector(plists), tfmda);
    search.initFullRange();
    EXPECT_TRUE(search.seek(1));
    EXPECT_TRUE(search.seek(4));
}

TEST(PredicateSearchTest, require_that_sorting_works_for_many_posting_lists) {
    MyPostingList plists[] = {{{1, 0x0001ffff},
                               {2, 0x0001ffff}},
                              {{2, 0x0001ffff},
                               {4, 0x0001ffff}},
                              {{2, 0x0001ffff},
                               {5, 0x0001ffff}},
                              {{2, 0x0001ffff},
                               {4, 0x0001ffff}},
                              {{2, 0x0001ffff},
                               {5, 0x0001ffff}}};
    MF mf{0, 1, 5, 0, 2, 2};
    CV cv{0, 1, 5, 0, 2, 2};
    IR ir(6, 0xffff);
    PredicateSearch search(&mf[0], &ir[0], 0xffff, cv, make_posting_lists_vector(plists), tfmda);
    search.initFullRange();
    EXPECT_TRUE(search.seek(1));
    EXPECT_TRUE(search.seek(2));
    EXPECT_TRUE(search.seek(4));
    EXPECT_TRUE(search.seek(5));
}

TEST(PredicateSearchTest, require_that_insufficient_interval_coverage_prevents_match) {
    MyPostingList plists[] = {{{2, 0x00010001},
                               {3, 0x0002ffff}}};
    MF mf{0, 0, 0, 0};
    CV cv{0, 0, 1, 1};
    IR ir(4, 0xffff);
    PredicateSearch search(&mf[0], &ir[0], 0xffff, cv, make_posting_lists_vector(plists), tfmda);
    search.initFullRange();
    EXPECT_FALSE(search.seek(2));
    EXPECT_FALSE(search.seek(3));
}

TEST(PredicateSearchTest, require_that_intervals_are_sorted) {
    MyPostingList plists[] = {{{2, 0x00010001}},
                              {{2, 0x0003ffff}},
                              {{2, 0x00020002}}};
    MF mf{0, 0, 0};
    CV cv{0, 0, 3};
    IR ir(3, 0xffff);
    PredicateSearch search(&mf[0], &ir[0], 0xffff, cv, make_posting_lists_vector(plists), tfmda);
    search.initFullRange();
    EXPECT_TRUE(search.seek(2));
}

TEST(PredicateSearchTest, require_that_NOT_is_supported_no_match) {
    MyPostingList plists[] = {{{2, 0x00010001}},  // [l, r]
                              {{2, 0x00010000},   // [l, r]*
                               {2, 0xffff0001}}};  // [r+1, r+1]*
    MF mf{0, 0, 0};
    CV cv{0, 0, 3};
    IR ir(3, 0xffff);
    PredicateSearch search(&mf[0], &ir[0], 0xffff, cv, make_posting_lists_vector(plists), tfmda);
    search.initFullRange();
    EXPECT_FALSE(search.seek(2));
}

TEST(PredicateSearchTest, require_that_NOT_is_supported_match) {
    MyPostingList plists[] = {{{2, 0x00010000},   // [l, r]*
                               {2, 0xffff0001}}};  // [r+1, r+1]*
    MF mf{0, 0, 0};
    CV cv{0, 0, 2};
    IR ir(3, 0xffff);
    PredicateSearch search(&mf[0], &ir[0], 0xffff, cv, make_posting_lists_vector(plists), tfmda);
    search.initFullRange();
    EXPECT_TRUE(search.seek(2));
}

TEST(PredicateSearchTest, require_that_NOT_is_supported_no_match_because_of_previous_term) {
    MyPostingList plists[] = {{{2, 0x00020001},   // [l, r]*
                               {2, 0xffff0002}}};  // [r+1, r+1]*
    MF mf{0, 0, 0};
    CV cv{0, 0, 2};
    IR ir(3, 0xffff);
    PredicateSearch search(&mf[0], &ir[0], 0xffff, cv, make_posting_lists_vector(plists), tfmda);
    search.initFullRange();
    EXPECT_FALSE(search.seek(2));
}

TEST(PredicateSearchTest, require_that_NOT_is_supported_subqueries) {
    MyPostingList plists[] = {{{2, 0x00010001}},  // [l, r]
                              {{2, 0x00010000},   // [l, r]*
                               {2, 0xffff0001}}};  // [r+1, r+1]*
    plists[0].setSubquery(0xffff);
    MF mf{0, 0, 0};
    CV cv{0, 0, 3};
    IR ir(3, 0xffff);
    PredicateSearch search(&mf[0], &ir[0], 0xffff, cv, make_posting_lists_vector(plists), tfmda);
    search.initFullRange();
    EXPECT_TRUE(search.seek(2));
}

TEST(PredicateSearchTest, require_that_there_can_be_many_intervals) {
    MyPostingList plists[] = {{{2, 0x00010001},
                               {2, 0x00020002},
                               {2, 0x00030003},
                               {2, 0x0001ffff},
                               {2, 0x00040004},
                               {2, 0x00050005},
                               {2, 0x00060006}}};
    MF mf{0, 0, 0};
    CV cv{0, 0, 7};
    IR ir(3, 0xffff);
    PredicateSearch search(&mf[0], &ir[0], 0xffff, cv, make_posting_lists_vector(plists), tfmda);
    search.initFullRange();
    EXPECT_TRUE(search.seek(2));
}

TEST(PredicateSearchTest, require_that_match_can_require_multiple_postinglists) {
    MyPostingList plists[] = {{{2, 0x00010001}},
                              {{2, 0x0002000b},
                               {2, 0x00030003}},
                              {{2, 0x00040003}},
                              {{2, 0x00050004}},
                              {{2, 0x00010008},
                               {2, 0x00060006}},
                              {{2, 0x00020002},
                               {2, 0x0007ffff}}};
    MF mf{0, 0, 0};
    CV cv{0, 0, 9};
    IR ir(3, 0xffff);
    PredicateSearch search(&mf[0], &ir[0], 0xffff, cv, make_posting_lists_vector(plists), tfmda);
    search.initFullRange();
    EXPECT_TRUE(search.seek(2));
}

TEST(PredicateSearchTest, require_that_subquery_bitmap_is_unpacked_to_subqueries) {
    MyPostingList plists[] = {{{2, 0x0001ffff}}};
    TermFieldMatchDataArray array;
    TermFieldMatchData data;
    array.add(&data);
    MF mf{0, 0, 0};
    CV cv{0, 0, 1};
    IR ir(3, 0xffff);
    PredicateSearch search(&mf[0], &ir[0], 0xffff, cv, make_posting_lists_vector(plists), array);
    search.initFullRange();
    EXPECT_TRUE(search.seek(2));
    search.unpack(2);
    EXPECT_EQ(0xffffffffffffffffUL, data.getSubqueries());
}


}  // namespace

GTEST_MAIN_RUN_ALL_TESTS()

File: predicate-search/src/test/java/com/yahoo/search/predicate/index/PredicateSearchTest.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.search.predicate.index;

import com.yahoo.search.predicate.Hit;
import com.yahoo.search.predicate.SubqueryBitmap;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * @author <a href="mailto:magnarn@yahoo-inc.com">Magnar Nedland</a>
 * @author bjorncs
 */
public class PredicateSearchTest {

    @Test
    void requireThatNoStreamsReturnNoResults() {
        PredicateSearch search = new PredicateSearch(new ArrayList<>(), new byte[0], new byte[0], new short[0], 1);
        assertEquals(0, search.stream().count());
    }

    @Test
    void requireThatSingleStreamFiltersOnConstructedCompleteIntervals() {
        PredicateSearch search = createPredicateSearch(
                new byte[]{1, 1, 1},
                postingList(
                        SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x000100ff),
                        entry(1, 0x00010001, 0x000200ff),
                        entry(2, 0x00010042)));
        assertEquals(Arrays.asList(new Hit(0), new Hit(1)).toString(), search.stream().toList().toString());
    }

    @Test
    void requireThatMinFeatureIsUsedToPruneResults() {
        PredicateSearch search = createPredicateSearch(
                new byte[]{3, 1},
                postingList(
                        SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x000100ff),
                        entry(1, 0x000100ff)));
        assertEquals(Arrays.asList(new Hit(1)).toString(), search.stream().toList().toString());
    }

    @Test
    void requireThatAHighKCanYieldResults() {
        PredicateSearch search = createPredicateSearch(
                new byte[]{2},
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x00010001)),
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x000200ff)));
        assertEquals(Arrays.asList(new Hit(0)).toString(), search.stream().toList().toString());
    }

    @Test
    void requireThatPostingListsAreSortedAfterAdvancing() {
        PredicateSearch search = createPredicateSearch(
                new byte[]{2, 1, 1, 1},
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x000100ff),
                        entry(3, 0x000100ff)),
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(1, 0x000100ff),
                        entry(2, 0x000100ff)));
        assertEquals(Arrays.asList(new Hit(1), new Hit(2), new Hit(3)).toString(), search.stream().toList().toString());
    }

    @Test
    void requireThatEmptyPostingListsWork() {
        PredicateSearch search = createPredicateSearch(
                new byte[0],
                postingList(SubqueryBitmap.ALL_SUBQUERIES));
        assertEquals(Arrays.asList().toString(), search.stream().toList().toString());
    }

    @Test
    void requireThatShorterPostingListEndingIsOk() {
        PredicateSearch search = createPredicateSearch(
                new byte[]{1, 1, 1},
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x000100ff),
                        entry(1, 0x000100ff)),
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(2, 0x000100ff)));
        assertEquals(Arrays.asList(new Hit(0), new Hit(1), new Hit(2)).toString(), search.stream().toList().toString());
    }

    @Test
    void requireThatSortingWorksForManyPostingLists() {
        PredicateSearch search = createPredicateSearch(
                new byte[]{1, 5, 2, 2},
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x000100ff),
                        entry(1, 0x000100ff)),
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(1, 0x000100ff),
                        entry(2, 0x000100ff)),
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(1, 0x000100ff),
                        entry(3, 0x000100ff)),
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(1, 0x000100ff),
                        entry(2, 0x000100ff)),
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(1, 0x000100ff),
                        entry(3, 0x000100ff)));
        assertEquals(
                Arrays.asList(new Hit(0), new Hit(1), new Hit(2), new Hit(3)).toString(),
                search.stream().toList().toString());
    }

    @Test
    void requireThatInsufficientIntervalCoveragePreventsMatch() {
        PredicateSearch search = createPredicateSearch(
                new byte[]{1, 1},
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x00010001),
                        entry(1, 0x000200ff)));
        assertEquals(Arrays.asList().toString(), search.stream().toList().toString());
    }

    @Test
    void requireThatIntervalsAreSorted() {
        PredicateSearch search = createPredicateSearch(
                new byte[]{1},
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x00010001)),
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x000300ff)),
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x00020002)));
        assertEquals(Arrays.asList(new Hit(0)).toString(), search.stream().toList().toString());
    }

    @Test
    void requireThatThereCanBeManyIntervals() {
        PredicateSearch search = createPredicateSearch(
                new byte[]{1},
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x00010001, 0x00020002, 0x00030003, 0x000100ff, 0x00040004, 0x00050005, 0x00060006)));
        assertEquals(Arrays.asList(new Hit(0)).toString(), search.stream().toList().toString());
    }

    @Test
    void requireThatNotIsSupported_NoMatch() {
        PredicateSearch search = createPredicateSearch(
                new byte[]{1},
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x00010001)),
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x00010000, 0x00ff0001)));
        assertEquals(Arrays.asList().toString(), search.stream().toList().toString());
    }

    @Test
    void requireThatNotIsSupported_Match() {
        PredicateSearch search = createPredicateSearch(
                new byte[]{1},
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x00010000, 0x00ff0001)));
        assertEquals(Arrays.asList(new Hit(0)).toString(), search.stream().toList().toString());
    }

    @Test
    void requireThatNotIsSupported_NoMatchBecauseOfPreviousTerm() {
        PredicateSearch search = createPredicateSearch(
                new byte[]{1},
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x00020001, 0x00ff0001)));
        assertEquals(Arrays.asList().toString(), search.stream().toList().toString());
    }

    @Test
    void requireThatIntervalSortingWorksAsUnsigned() {
        PredicateSearch search = createPredicateSearch(
                new byte[]{1},
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x00010001)),
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x00fe0001, 0x00ff00fe)));
        assertEquals(Arrays.asList(new Hit(0)).toString(), search.stream().toList().toString());
    }

    @Test
    void requireThatMatchCanRequireMultiplePostingLists() {
        PredicateSearch search = createPredicateSearch(
                new byte[]{6},
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x00010001)),
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x0002000b, 0x00030003)),
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x00040003)),
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x00050004)),
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x00010008, 0x00060006)),
                postingList(SubqueryBitmap.ALL_SUBQUERIES,
                        entry(0, 0x00020002, 0x000700ff)));
        assertEquals(Arrays.asList(new Hit(0)).toString(), search.stream().toList().toString());
    }

    private static PredicateSearch createPredicateSearch(byte[] minFeatures, PostingList... postingLists) {
        byte[] nPostingListsForDocument = new byte[minFeatures.length];
        short[] intervalEnds = new short[minFeatures.length];
        Arrays.fill(intervalEnds, (short) 0xFF);
        List<PostingList> list = Arrays.asList(postingLists);
        for (PostingList postingList : postingLists) {
            for (int id : postingList.getDocIds()) {
                nPostingListsForDocument[id]++;
            }
        }
        return new PredicateSearch(list, nPostingListsForDocument, minFeatures, intervalEnds, 0xFF);
    }

    private static class SimplePostingList implements PostingList {
        private final long subquery;
        private final Entry[] entries;
        private int[] currentIntervals;
        private int currentIntervalIndex;
        private int currentDocId;
        private int currentIndex;

        public SimplePostingList(long subquery, Entry... entries) {
            this.subquery = subquery;
            this.entries = entries;
            this.currentIndex = 0;
            this.currentDocId = -1;
        }

        @Override
        public boolean nextDocument(int docId) {
            while (currentIndex < entries.length && entries[currentIndex].docId <= docId) {
                ++currentIndex;
            }
            if (currentIndex == entries.length) {
                return false;
            }
            Entry entry = entries[currentIndex];
            currentDocId = entry.docId;
            currentIntervals = entry.intervals;
            currentIntervalIndex = 0;
            return true;
        }

        @Override
        public boolean prepareIntervals() {
            return true;
        }

        @Override
        public boolean nextInterval() {
            return ++currentIntervalIndex < currentIntervals.length;
        }

        @Override
        public int getDocId() {
            return currentDocId;
        }

        @Override
        public int size() {
            return entries.length;
        }

        @Override
        public int getInterval() {
            return currentIntervals[currentIntervalIndex];
        }

        @Override
        public long getSubquery() {
            return subquery;
        }

        @Override
        public int[] getDocIds() {
            return Arrays.stream(entries).mapToInt(e -> e.docId).toArray();
        }

        public static class Entry {
            public final int docId;
            public final int[] intervals;

            public Entry(int docId, int... intervals) {
                this.docId = docId;
                this.intervals = intervals;
            }
        }
    }

    private static SimplePostingList postingList(long subquery, SimplePostingList.Entry... entries) {
        return new SimplePostingList(subquery, entries);
    }

    private static SimplePostingList.Entry entry(int docId, int... intervals) {
        return new SimplePostingList.Entry(docId, intervals);
    }
}
